### ISSUE TITLE
Data engine

### ISSUE THREAD

**Issue by gaow** [vatlab/VAT#3]
I wonder if we can start brainstorming about possible data engines. I'm currently using HDF5 mostly because my colleagues use R and matlab not Python. HDF5 does decent job in structured storage, query and above all can be easily shared with R / matlab users. There are a fair amount of negative reviews about it and I am not 100% happy with it either. If we can start looking for some data engine I can start to develop my project using the same backbone, which makes possible merger easier.

**Comment by gaow** [comment]
Chen 2016 uses NoSQL. A google search of NoSQL + genomics returns a lot relevant developments

**Comment by BoPeng** [comment]
NoSQL is certainly a possibility. It is notoriously difficult to set up and use but it scales well, with cloud support and allows the hadoop/spark style of data processing. The problem is that Chen 2016 is providing a service that does not need to move data around. This is not the model we will use if we take user input and process them because deploying user data is very tedious.

**Comment by BoPeng** [comment]
I would like to add that one major advantage of HDF format is that it allows parallel read/write, which could be very help for data analysis.

**Comment by gaow** [comment]
Yes, providing it as a service removes the pain on the user's end to set it up. But that can be a big issue for our application! The target of VAT was small labs that just needs to get things done without a fancy computational infrastructure. HDF5 may fit better for that. Although we should also see if it is possible to get it work on modern distributed computational environment.

One thing I dislike about HDF5 is that neither of the two Python API is satisfactory. `h5py` is more native thus has features such as parallel I/O but `pytables` is more database like thus perhaps easier to use. We originally proposed some hybrid model but I'm still not comfortable with anything involving `sqlite`, to store over 100 million variants.

**Comment by BoPeng** [comment]
Given the trouble we had with the flexibility to handle all sort of input data, I tend to think we should only have native support for VCF. We can keep the .fmt feature for annotation databases because they have diverse format in nature, but for raw variant and genotype data, it should be ok to accept only VCF format (and provide converter script if needed). 

We then can work with hdf5 C API directly and do not need any python wrapper.

The bottom line is that we can sacrifice some flexibility for performance when we are dealing with large datasets.

**Comment by gaow** [comment]
Yes, and also BCF files. For variant data it is often trivial to convert to VCF. For data with sample genotypes the only other format we support was PLINK binary and we support it essentially by converting it to VCF-like, so we can drop that, too. 

Not sure why we want to work at C level, though (unless we want to go directly from BCF to HDF5??). `h5py` has a low level API in Cython that provides access to nearly all of the HDF5 C API as it claims. At least we want to leverage `h5py` native numpy data type. Still, queries with HDF5 is hard unless with `pytable`

**Comment by BoPeng** [comment]
We do not have to work at the C level, but if there is no need to reuse the code, we can write the import vcf part completely in C to achieve maximum performance. The query part can use python though.

**Comment by gaow** [comment]
That makes sense. But before devoting too much into importing, we probably should test out the entire design and see how easy/fast it is to query -- just in case we imported all the data but we do not like the organization of the data. What's the best way to implement a test benchmark to explore data storage? 

For example, we want to compare designs where we store batches of samples in different HDF5 vs batches of variants in HDF5, or both ... Also for expression data we'll have over 20,000 (eQTL) times N phenotypes for each sample, where N is the number of tissues / cells measured on that sample. So perhaps the first thing to do is to design and figure out how to test our designs rather than worry about import.

**Comment by BoPeng** [comment]
Of course the implementation would be based on extensive testing and comparison...  I plan to use complete 1000 genome dataset as test data for importing and query. We can even simulate more data based on 1000g.

**Comment by gaow** [comment]
Should we dare ourselves to start 1KG as a toy and UK10K the real deal? UK100K is on the way ... but it should be good enough to say we can handle UK10K!

**Comment by BoPeng** [comment]
If we can handle UK10K, we should be able to handle 100K. I mean, with the introduction of unified variant ID, we can separate variants and genotypes into (numerous) small projects. They can be imported separately and in parallel (which should have acceptable performance on a network filesystem such as GPFS that is designed for parallel IO), and be analyzed only when needed.

We perhaps still need a "project" to store variant tables etc but the actual variants and genotypes will be stored in different files, perhaps even in different formats.

**Comment by BoPeng** [comment]
I suppose we can implement a storage-neutral solution here. Conceptually speaking, we can implement a protocol that is used by the vat program to query data from different sources. Then we can implement a few "data providers" that provide data through this interface. For example,

1. we can provide a "native vcf" data provider that query data directly from vcf files.
2. we can provide a "hdf5" data provider that query data from hdf5 format.
3. we can provide a hadoop data provider that query data stored in HDFS storage.
4. we can provide a sqlite3 data provider that query data from existing variant tools projects.
5. we can provide a network interface to query data that are store from another server.

The data providers are separate daemon processes for each datasource. They are responsible for digesting input data files, process and store them (e.g. compress and index vcf), and provide data to VAT in a most efficient manner. The VAT process basically coordinate the association analysis process ... If and how exactly the data will be transfer to vat needs to be determined later. 

The advantages of this approach include

1. It allows distributed and parallel processing of data. For example, we can separate large datasets by chromosome or by population and use multiple providers to provide data. 
2. We can implement and improve data providers separately, and leave room for future development.
3. This allows the development and test of association analysis methods without waiting for the maturity of all data providers. As soon as we develop a basic data model (e.g. for vcf), we can work on all sort of statistical methods.

Potential problems include:

1. We might have to implement a common set of operations for data providers, this can potentially be achieved by providing a common library to these providers.
2. It is unclear how much the overhead of this approach is.

**Comment by gaow** [comment]
I want to clarify 2 questions here: 

1. I'm worried about query and I'm not sure how this approach works on queries efficiently. For vtools 2X (I'll call it version 2X assuming the new project is version 3X under the name VAT) queries are a big part of it. We customized many sqlite3 functions so that we can utilize sqlite3 in place without having to output data. I was hoping something similar be down with pytables on hdf5. With the idea of data provider, it looks like any operation on data will involve dumping slices of data and feeding to computational routines. Is that the case? 

2. How do we store summary statistics computed from data? I assume we use separate files which are called a project, collectively? 

My understanding is that we can accept native VCF to enter the association analysis step directly. But we'll not provide advanced queries / data quality control for VCF files because those should be done in hdf5 or sqlite3 format.

Being able to use hadoop is definitely appealing. There is a discussion on this from the hdf5 team:

https://www.hdfgroup.org/pubs/papers/Big_HDF_FAQs.pdf

Getting hadoop work with other data formats might be harder.

**Comment by gaow** [comment]
It might worth to have a chat with someone, for example Fuli Yu, at Baylor HGSC. I heard they use MapReduce in production.

**Comment by BoPeng** [comment]
MapReduce can certainly be used for association analysis, but we would like to be able to handle different types of data (projects).

**Comment by BoPeng** [comment]
When I talked about data-neutral storage and data providers, I was thinking of socket-programming. Namely each data provider provides data through a socket, regardless where the data provider is (local or remote) and in which format the data is actually stored (vcf, hdf5 etc). 

When we add a data source to a project (analysis), vat basically creates a data provider or connect to an existing (remote) data provider. It then asks the data providers what types of data they provide (samples, variants, genotypes, summary statistics etc), decide what data it needs for the particular association analysis,  query the data provider for data, and dispatch data to workers. This process is complex and will be slower for small datasets, but can hopefully be faster in large scale projects.

The data provider is more or less independent of the vat process. We proposed in our grant application to provide web based analysis for which users might fire multiple vat instances that reuse the same set of (reference) data. Instead of creating multiple processes to access the same data (could be source of race condition etc), we can let one data provider respond to multiple requests, which could be made efficiency if the provider can make use of cache etc.

I do not know how to achieve this in an efficient way but we could exercise caching and asynchronization in this model. Basically vat asks for all data (not for example a gene) from all providers to reduce socket overhead, and the provider can cache data if same blocks of data was asked before.
